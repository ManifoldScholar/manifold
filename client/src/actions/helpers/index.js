export { default as createApiAction } from "./createApiAction";
export { default as commonActions } from "./commonActions";
