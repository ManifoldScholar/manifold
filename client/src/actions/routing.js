import { createAction } from "redux-actions";

export const update = createAction("ROUTE_UPDATE", state => state);
