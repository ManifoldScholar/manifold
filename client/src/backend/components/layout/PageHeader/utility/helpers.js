import lh from "helpers/linkHandler";

export const getTextLinks = ({ texts, pathname }) =>
  texts?.length
    ? texts?.map(t => ({
        label: t.label,
        route: "backendTextAnalytics",
        id: t.id,
        active: pathname?.includes(t.id)
      }))
    : null;

export const getIssueLinks = ({ issues, pathname }) =>
  issues?.length
    ? issues?.map(i => ({
        label: i.label,
        route: "backendProjectAnalytics",
        id: i.id,
        active: pathname?.includes(i.id)
      }))
    : null;

export const getLinkOrButtonProps = action => {
  if (action.href) {
    return action.download
      ? { as: "a", href: action.href, download: true }
      : {
          as: "a",
          href: action.href,
          target: "_blank",
          re: "noopener noreferrer"
        };
  }

  if (action.route)
    return {
      to: lh.link(action.route, action.slug, action.resourceSlug)
    };

  return { as: "button", onClick: action.onClick };
};
