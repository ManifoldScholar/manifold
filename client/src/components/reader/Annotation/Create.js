import React, { PureComponent, PropTypes } from 'react';
import { Annotation } from 'components/reader';

export default class AnnotationCreate extends PureComponent {

  static displayName = "Annotation.Create";

  static propTypes = {
    annotation: PropTypes.object,
    selection: PropTypes.object,
    closeDrawer: PropTypes.func
  }

  render() {
    return (
      <Annotation.Selection.Wrapper {...this.props} annotating />
    );
  }

}
