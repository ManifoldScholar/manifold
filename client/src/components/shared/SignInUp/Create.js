import React, { Component, PropTypes } from 'react';
import usersAPI from '../../../api/users';
import { request, flush } from '../../../actions/shared/entityStore';
import { select } from '../../../utils/entityUtils';
import { Errorable } from '../Form';
import { connect } from 'react-redux';
import get from 'lodash/get';
import { startLogin } from '../../../actions/shared/authentication';

class Create extends Component {

  static propTypes = {
    dispatch: PropTypes.func,
    response: PropTypes.object,
    user: PropTypes.object,
    showForgot: PropTypes.func.isRequired,
    showLogin: PropTypes.func.isRequired,
    showUpdate: PropTypes.func.isRequired
  };

  static requests = {
    create: 'signinup-create-user'
  };

  static mapStateToProps(state) {
    const r = Create.requests;
    const myState = {
      user: select(r.create, state.entityStore),
      response: state.entityStore.responses[r.create]
    };
    return myState;
  }

  constructor() {
    super();
    this.createUser = this.createUser.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.authenticateUser = this.authenticateUser.bind(this);
    this.state = {
      authenticating: false,
      user: {
        email: `zach+${new Date().getMilliseconds()}@castironcoding.com`,
        name: 'Zach Davis',
        password: 'testtest123',
        passwordConfirmation: 'testtest123'
      }
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.user && !this.props.user) {
      this.authenticateUser();
    }
    if (nextProps.authentication.authenticated) {
      this.props.showUpdate();
    }
  }

  componentWillUnmount() {
    this.props.dispatch(flush(Create.requests));
  }

  authenticateUser() {
    this.setState({ authenticating: true });
    const { dispatch } = this.props;
    dispatch(startLogin(this.state.user.email, this.state.user.password));
  }

  createUser(event) {
    event.preventDefault(event.target);
    this.props.dispatch(request(usersAPI.create(this.state.user), Create.requests.create));
  }

  handleInputChange(event) {
    const user =
      Object.assign({}, this.state.user, { [event.target.name]: event.target.value });
    this.setState({ user });
  }

  render() {
    let errors = get(this.props.response, 'errors') || {};
    return (
      <div>
        <form method="post" onSubmit={this.createUser} >
          <h4 className="form-heading">Create Account</h4>
          <div className="row-1-p">
            <Errorable className="form-input" field="email" errors={errors} >
             <label>Email</label>
              <input
                value={this.state.user.email}
                type="text"
                name="email"
                id="create-email"
                onChange={this.handleInputChange}
                placeholder="Email"
              />
            </Errorable>
          </div>
          <div className="row-1-p">
            <Errorable
              className="form-input"
              field={['first_name', 'last_name']}
              errors={errors}
            >
              <label>
                Name
              </label>
              <input
                value={this.state.user.name}
                type="text"
                id="create-name"
                name="name"
                onChange={this.handleInputChange}
                placeholder="Name"
              />
            </Errorable>
          </div>
          <div className="row-1-p">
            <Errorable className="form-input" field="password" errors={errors} >
              <label>
                Password
              </label>
              <input
                value={this.state.user.password}
                type="password"
                name="password"
                id="create-password"
                onChange={this.handleInputChange}
                placeholder="password"
              />
            </Errorable>
          </div>
          <div className="row-1-p">
            <div className="form-input">
              <label>
                Confirm Password
              </label>
              <input
                value={this.state.user.passwordConfirmation}
                type="password"
                name="passwordConfirmation"
                id="create-password-confirmation"
                onChange={this.handleInputChange}
                placeholder="Confirm Password"
              />
            </div>
          </div>
          <div className="row-1-p">
            <div className="form-input">
              <input
                className="button-secondary button-with-room"
                type="submit"
                value="Create Account"
              />
            </div>
          </div>
        </form>
        <p className="login-links">
          {'You can also create a Manifold account using your Facebook or Twitter credentials.'}
        </p>
        <section className="login-external">
          <button className="button-secondary-dark">
            <i className="manicon manicon-facebook"></i>
            <span>Log in with Facebook</span>
          </button>
          <button className="button-secondary-dark">
            <i className="manicon manicon-twitter"></i>
            <span>Log in with Twitter</span>
          </button>
        </section>
        <p className="login-links">
          <a href="#" onClick={this.props.showLogin}>
            {'Already have an account?'}
          </a>
        </p>
        <p className="login-links">
          {'By creating this account, you agree to Manifold\'s terms and conditions.'}
        </p>

      </div>
    );
  }
}

const SignUpCreate = connect(
  Create.mapStateToProps
)(Create);
export default SignUpCreate;
