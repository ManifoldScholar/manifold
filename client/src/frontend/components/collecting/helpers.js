import flatMapDepth from "lodash/flatMapDepth";
import identity from "lodash/identity";
import has from "lodash/has";
import get from "lodash/get";
import isEmpty from "lodash/isEmpty";

export function getEntityCollection(entity, relationship = "collection") {
  return get(entity, `relationships.${relationship}`);
}

function isCollection(possibleCollection) {
  return has(possibleCollection, "attributes.categoryMappings");
}

function hasCollection(possibleEntity, relationship = "collection") {
  return has(possibleEntity, `relationships.${relationship}`);
}

export function collectedIdsForCollection(collection) {
  if (!collection) return [];
  const mappings = collection.attributes.categoryMappings;
  return flatMapDepth(
    Object.values(mappings).map(mapping => Object.values(mapping)),
    identity,
    2
  );
}

export function inCollection(collectionOrEntity, collectable) {
  const collection = isCollection(collectionOrEntity)
    ? collectionOrEntity
    : hasCollection(collectionOrEntity) &&
      getEntityCollection(collectionOrEntity);
  if (!collection) return false;

  const collectedIds = collectedIdsForCollection(collection);
  return collectedIds.includes(collectable.id);
}

export function inCollections(collectable, ...collectionsOrEntities) {
  return collectionsOrEntities.some(collectionOrEntity =>
    inCollection(collectionOrEntity, collectable)
  );
}

export function getResponse(id, responses) {
  return responses.find(res => res.id === id);
}

export function idInResponses(id, responses) {
  if (responses?.length < 1) return false;
  return responses.map(res => res.id).includes(id);
}

export function hasItemsInCollection(entity) {
  const collection = getEntityCollection(entity);
  if (!collection?.attributes) return false;
  return !isEmpty(collection.attributes.categoryMappings);
}
