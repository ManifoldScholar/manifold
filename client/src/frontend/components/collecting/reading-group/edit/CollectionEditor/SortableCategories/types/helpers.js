import PropTypes from "prop-types";

export const collectedShape = {
  categoryId: PropTypes.string.isRequired,
  collectedIds: PropTypes.array.isRequired,
  responses: PropTypes.array.isRequired,
  onRemove: PropTypes.func.isRequired,
  showDropzone: PropTypes.bool
};
