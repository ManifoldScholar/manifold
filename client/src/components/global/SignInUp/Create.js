import React, { Component, PropTypes } from 'react';
import { usersAPI, requests } from 'api';
import { entityStoreActions, currentUserActions } from 'actions';
import { select } from 'utils/entityUtils';
import { Form, SignInUp } from 'components/global';
import connectAndFetch from 'utils/connectAndFetch';
import get from 'lodash/get';

const { request, flush } = entityStoreActions;

class CreateContainer extends Component {

  static propTypes = {
    dispatch: PropTypes.func,
    response: PropTypes.object,
    user: PropTypes.object,
    showForgot: PropTypes.func.isRequired,
    showLogin: PropTypes.func.isRequired,
    showCreateUpdate: PropTypes.func.isRequired
  };

  static mapStateToProps(state) {
    const myState = {
      user: select(requests.gCreateUser, state.entityStore),
      response: state.entityStore.responses[requests.gCreateUser]
    };
    return myState;
  }

  constructor() {
    super();
    this.createUser = this.createUser.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
    this.authenticateUser = this.authenticateUser.bind(this);
    this.state = {
      authenticating: false,
      user: {
        email: '',
        name: '',
        password: '',
        passwordConfirmation: ''
      }
    };
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.user && !this.props.user) {
      this.authenticateUser();
    }
    if (nextProps.authentication.authenticated) {
      this.props.showCreateUpdate();
    }
  }

  componentWillUnmount() {
    this.props.dispatch(flush(requests.gCreateUser));
  }

  authenticateUser() {
    this.setState({ authenticating: true });
    const { dispatch } = this.props;
    dispatch(currentUserActions.login({
      email: this.state.user.email,
      password: this.state.user.password
    }));
  }

  createUser(event) {
    event.preventDefault(event.target);
    this.props.dispatch(request(usersAPI.create(this.state.user), requests.gCreateUser));
  }

  handleInputChange(event) {
    const user =
      Object.assign({}, this.state.user, { [event.target.name]: event.target.value });
    this.setState({ user });
  }

  render() {
    let errors = get(this.props.response, 'errors') || [];
    return (
      <div>
        <form method="post" onSubmit={this.createUser} >
          <h4 className="form-heading">Create Account</h4>
          <div className="row-1-p">
            <Form.Errorable
              className="form-input"
              name="attributes[email]"
              errors={errors}
            >
             <label>Email</label>
              <input
                value={this.state.user.email}
                type="text"
                name="email"
                id="create-email"
                onChange={this.handleInputChange}
                placeholder="Email"
              />
            </Form.Errorable>
          </div>
          <div className="row-1-p">
            <Form.Errorable
              className="form-input"
              name={['attributes[firstName]', 'attributes[lastName]']}
              errors={errors}
            >
              <label>
                Name
              </label>
              <input
                value={this.state.user.name}
                type="text"
                id="create-name"
                name="name"
                onChange={this.handleInputChange}
                placeholder="Name"
              />
            </Form.Errorable>
          </div>
          <div className="row-1-p">
            <Form.Errorable
              className="form-input"
              name="attributes[password]"
              errors={errors}
            >
              <label>
                Password
              </label>
              <input
                value={this.state.user.password}
                type="password"
                name="password"
                id="create-password"
                onChange={this.handleInputChange}
                placeholder="Password"
              />
            </Form.Errorable>
          </div>
          <div className="row-1-p">
            <Form.Errorable
              className="form-input"
              name="attributes[passwordConfirmation]"
              errors={errors}
            >
              <label>
                Confirm Password
              </label>
              <input
                value={this.state.user.passwordConfirmation}
                type="password"
                name="passwordConfirmation"
                id="create-password-confirmation"
                onChange={this.handleInputChange}
                placeholder="Confirm Password"
              />
            </Form.Errorable>
          </div>
          <div className="row-1-p">
            <div className="form-input">
              <input
                className="button-secondary button-with-room"
                type="submit"
                value="Create Account"
              />
            </div>
          </div>
        </form>
        <p className="login-links">
          {'You can also create a Manifold account using your Facebook, Twitter, or ' +
          'Google credentials.'}
        </p>
        <section className="login-external">
          <SignInUp.Oauth.Monitor dispatch={this.props.dispatch} />
          <SignInUp.Oauth.Button
            dispatch={this.props.dispatch}
            provider="facebook"
          >
            <span>Log in with Facebook</span>
          </SignInUp.Oauth.Button>
          <SignInUp.Oauth.Button
            dispatch={this.props.dispatch}
            provider="google"
            iconName="manicon-envelope"
          >
            <span>Log in with Google</span>
          </SignInUp.Oauth.Button>
          <SignInUp.Oauth.Button
            dispatch={this.props.dispatch}
            provider="twitter"
          >
            <span>Log in with Twitter</span>
          </SignInUp.Oauth.Button>
        </section>
        <p className="login-links">
          {'By creating this account, you agree to Manifold\'s terms and conditions.'}
        </p>
        <p className="login-links">
          <a href="#" onClick={this.props.showLogin}>
            {'Already have an account?'}
          </a>
        </p>
      </div>
    );
  }
}

export default connectAndFetch(CreateContainer);
