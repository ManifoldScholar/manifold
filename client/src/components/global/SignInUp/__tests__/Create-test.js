import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import Create from '../Create';
import build from 'test/fixtures/build';
import { Provider } from 'react-redux';
import { wrapWithRouter, renderWithRouter } from 'test/helpers/routing';

describe("Global.SignInUp.Create component", () => {

  const store = build.store();

  const showForgot = jest.fn();
  const showLogin = jest.fn();
  const showCreateUpdate = jest.fn();
  const user = build.entity.user("1");

  const root = (
    wrapWithRouter(
      <Provider store={store} >
        <Create
          dispatch={store.dispatch}
          showForgot={showForgot}
          showLogin={showLogin}
          showCreateUpdate={showCreateUpdate}
          user={user}
        />
      </Provider>
    )
  );

  it('renders correctly', () => {
    const component = renderer.create(root);
    let tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should trigger showLogin callback when show login is clicked', () => {
    const wrapper = mount(wrapWithRouter(root));
    showLogin.mockClear();
    wrapper.find('[data-id="show-login"]').first().simulate('click');
    expect(showLogin).toHaveBeenCalled();
  });
});
