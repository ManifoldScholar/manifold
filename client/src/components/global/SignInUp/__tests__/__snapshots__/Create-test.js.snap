// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Global.SignInUp.Create component renders correctly 1`] = `
<div>
  <form
    method="post"
    onSubmit={[Function]}
  >
    <h4
      className="form-heading"
    >
      Create Account
    </h4>
    <div
      className="row-1-p"
    >
      <div
        className="form-input"
        style={Object {}}
      >
        <label>
          Email
        </label>
        <input
          id="create-email"
          name="email"
          onChange={[Function]}
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="row-1-p"
    >
      <div
        className="form-input"
        style={Object {}}
      >
        <label>
          Name
        </label>
        <input
          id="create-name"
          name="name"
          onChange={[Function]}
          placeholder="Name"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      className="row-1-p"
    >
      <div
        className="form-input"
        style={Object {}}
      >
        <label>
          Password
        </label>
        <input
          id="create-password"
          name="password"
          onChange={[Function]}
          placeholder="Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="row-1-p"
    >
      <div
        className="form-input"
        style={Object {}}
      >
        <label>
          Confirm Password
        </label>
        <input
          id="create-password-confirmation"
          name="passwordConfirmation"
          onChange={[Function]}
          placeholder="Confirm Password"
          type="password"
          value=""
        />
      </div>
    </div>
    <div
      className="row-1-p"
    >
      <div
        className="form-input"
      >
        <input
          className="button-secondary button-with-room"
          type="submit"
          value="Create Account"
        />
      </div>
    </div>
  </form>
  <p
    className="login-links"
  >
    You can also create a Manifold account using your Facebook, Twitter, or Google credentials.
  </p>
  <section
    className="login-external"
  >
    <div />
    <button
      className="button-secondary-dark"
      onClick={[Function]}
    >
      <i
        className="manicon manicon-envelope"
      />
      <span>
        Log in with Google
      </span>
    </button>
    <button
      className="button-secondary-dark"
      onClick={[Function]}
    >
      <i
        className="manicon manicon-twitter"
      />
      <span>
        Log in with Twitter
      </span>
    </button>
  </section>
  <p
    className="login-links"
  >
    By creating this account, you agree to Manifold's terms and conditions.
  </p>
  <p
    className="login-links"
  >
    <a
      data-id="show-login"
      href="#"
      onClick={[Function]}
    >
      Already have an account?
    </a>
  </p>
</div>
`;
